### ISSUE TITLE
Deploy workflows as DataFlow templates

### ISSUE THREAD

**Issue by cwbeitel** [iqtk/iqtk#3]
DataFlow now supports templates via Python. Also you can trigger templates over HTTP from a cloud function as well as from the DataFlow UI. Which means that if instead of deploying the whole toolkit for end users to use to run workflows if they are deployed as templates they can just use those from the DataFlow UI. And instead of running and managing containers with the library for submitting DataFlow jobs those jobs can just be triggered directly from the cloud function we have running that responds to storage events without the rest of the complexity.

From the [DataFlow template docs](https://cloud.google.com/dataflow/docs/templates/executing-templates), submission from the DataFlow UI would be similar to the following:

![](https://cloud.google.com/dataflow/images/wordcount_template_execute.png)

Triggering a template job over http, such as from [our cloud function](https://github.com/iqtk/iqtk/blob/master/inquiry/functions/storage-trigger/index.js) that responds to storage events.

```bash
    POST https://dataflow.googleapis.com/v1b3/projects/[YOUR_PROJECT_ID]/templates
    {
        "jobName": "[JOB_NAME]",
        "gcsPath":"gs://[YOUR_BUCKET_NAME]/templates/TemplateName",
        "parameters": {
            "topic": "projects/[YOUR_PROJECT_ID]/topics/[YOUR_TOPIC_NAME]",
            "table": "[YOUR_PROJECT_ID]:[YOUR_DATASET].[YOUR_TABLE_NAME]"
        },
        "environment": {
            "tempLocation": "gs://[YOUR_BUCKET_NAME]",
            "zone": "us-central1-f"
        }
    }
```

Doing it this way in the short term would mean that that function wouldn't send out messages to PubSub to be received by the rest of some complex infrastructure, the entire deployment would consist of that single cloud function and templates in a GCS bucket. I.e. not only would we have removed the operational burden and complexity of managing an AirFlow deployment (by being able to submit jobs to the DataFlow service) we would be able to make these workflows available for anyone in the world to run with zero install from an already existing UI for both submission and monitoring.

**Comment by cwbeitel** [comment]
Planning to revise Workflow subclass init pattern to be the following

```python
class SimpleWorkflow(Workflow):
    """Provides a simple illustration of workflow structure."""

    def __init__(self):
        self.meta = {
            "name": "rna_quantification",
            "description": "Use the tuxedo transcriptome analysis suite to obtain differential transcript expression levels for an RNA-seq dataset.",
            "parameters": [{
                "name": "config_file_path",
                "label": "Config path",
                "help_text": "GCS path to run config file.",
                "regexes": ["^gs:\/\/[^\n\r]+$"],
                "is_optional": False
                }]
            }
        super(SimpleWorkflow, self).__init__()

```
mirroring the syntax of the DataFlow template metadata file. Replacing the use of self.arg_template

```python
class SimpleWorkflow(Workflow):
    """Provides a simple illustration of workflow structure."""

    def __init__(self):
        self.tag = 'simple-workflow'
        self.arg_template = {
            'file_path': {
                'help': 'path to a gcs file'
                },
            'scalar_param': {
                'help': 'some scalar'
                }
            }
        super(SimpleWorkflow, self).__init__()
```

for the purpose of being able to generate the metadata file needed by dataflow to know what fields to present to user, per [dflow template instructions](https://cloud.google.com/dataflow/docs/templates/creating-templates#metadata-parameters). When a run is specified to be a template upload instead of a regular run, this will be used to generate a [workflow_name]_metadata file that will be uploaded to the same GCS dir that is specified to be the target path for the template upload. (It appears that by specifying a --template_location arg the request to dataflow is interpreted as a template upload only and not run.)

`Workflow.meta` will then be used in a similar way to how `arg_template` is used now to loop through the template and `add_argument` to a `parser` instance except this will be done in the `_add_argparse_args` method of a `PipelineOptions` subclass more or less as 

```python
def user_opts_factory(opt_template):
    class UserOptions(PipelineOptions):
        @classmethod
        def _add_argparse_args(cls, parser):
            for opt in opt_template:
                # parse opt into the_tag, the_type, etc.
                parser.add_value_provider_argument('--%s' % the_tag, type=the_type)
    return UserOptions
```

Also we'll need to choose between `parser.add_value_provider_argument` and `parser.add_argument` depending on whether the run is a template upload or a regular run.

Of course one question on my mind is if there's a simpler way to do that.

**Comment by jlewi** [comment]
Will the limitations of Dataflow Templates be an issue? For example, I don't think you can change the graph. So suppose a pipeline can accept different file formats and based on the file format you need to run  different transforms to produce a common datatype that will then be processed by a common set of transforms. I don't think you can do this with templates.

If you want to create a REST API to trigger a Dataflow why not just put your Dataflow program in a container and then trigger execution of the container? For example, if you have a K8s cluster you could use a custom resource or a framework like [fission.io](http://fission.io/).

**Comment by cwbeitel** [comment]
Good point. We won't be able to change the graph. I was expecting to only accept a static set of formats except allowing them to be compressed in various forms. In [`toolkit.metabolomics.operations.MSConvert`](https://github.com/iqtk/iqtk/blob/master/inquiry/toolkit/metabolomics/operations.py) it checks for whether the file ends in `.tar.gz` or `.tgz` and optionally prepends a decompression step to the command its constructing, 

```python
class MSConvert(iqf.task.ContainerTask):

    def __init__(self, args):
        """Initialize an MSConvert container task."""
        container = iqf.task.ContainerTaskResources(
            disk=60, cpu_cores=4, ram=8,
            image='gcr.io/jbei-cloud/msconvert:0.0.1')
        super(MSConvert, self).__init__(task_label='msconvert', args=args,
                                        container=container)

    def process(self, file_path):

        # Declare intermediates
        decomp_target = self.out_path + '/decompressed'
        converted_target = localize('converted-%s.mzML' % iqf.util.shaker.shake(),
                                    self.out_path)

        # Construct command
        cmd = iqf.util.Command(['mkdir', '-p', decomp_target])

        if file_path.endswith('.tar.gz') or file_path.endswith('.tgz'):
            cmd.chain(['tar', '-xzf', localize(file_path),
                       '-C', decomp_target, '--strip-components=1'])

        cmd.chain(["wine", "msconvert", decomp_target, "--zlib",
                           "-o", self.out_path])
        cmd.chain(["mv", self.out_path + "/*.mzML", converted_target])

        yield self.submit(cmd.txt, inputs=[file_path],
                          expected_outputs=[{'name': 'converted.mzML',
                                             'file_type': 'mzml'}])

```

You could use this pattern in the first step of a pipeline meant to accept a range of input types. But yeah if you want to do things like conditionally run the Bowtie or BWA DoFn based on a runtime param that's probably not an easy thing to do. One option would be to have a general DoFn and treat the `BWA`, `Bowtie`, `TopHat`, etc. operators as pure functions that are selected at runtime. But yeah generally I agree it's not clear how you would hack in the ability to re-wire graphs and using DataFlow introduces unnecessary complexity into the development experience in exchange for the short-term win that anyone anywhere would be able to run templates without having to deploy or manage anything, from the dflow UI.

But I've also been putting together the fission and straight kube approaches in parallel ([experiments](https://github.com/iqtk/iqtk/tree/e0a404206abb030c30384f2e3d63defe98930357/inquiry/tools)) and it should be easy enough to deploy this in that way as well in the interest in moving in the direction of a more general ecosystem of components and wiring than is suitable for DataFlow.

**Comment by jlewi** [comment]
Makes sense thanks for the explanation.

**Comment by cwbeitel** [comment]
Yeah there's still a lot to consider there. If a novice gcp user provides a file with the right regex pattern but just isn't the right underlying format or mis-parameterizes in some other way that passes the initial regex (or for whatever other reason) and a pipeline fails they'll be at a loss for what to do unless they can find and interpret the logs..... I think it's over-extending to try for that right at this moment and might be a fair amount more usable with a custom UI, more extensive pre-validation than you can do with regex's, and user-appropriate error messages instead of raw logs.
